[android] Set decimal separator depending on locale

gpesquero opened the pull request:
Signed-off-by: Gonzalo Pesquero <gpesquero@yahoo.es>

Proposal for setting the decimal separator based on system locale for distances in some controls and text views of the app in Android, as suggested in [issue 3992](https://github.com/organicmaps/organicmaps/issues/3992).

Checking the code, I've noticed that all the displayed distances come as strings from jni calls, so the proposed solution just replaces the decimal dot '.' with the system locale separator (commonly a comma ',') when applicable. 

This is just a first approach for only a few controls and panels and it has been correctly tested in a Android simulator with different system languages.

The proposed approach may not be the best solution, so any comment or feedback is welcomed. This PR may be completed later to implement the remaining controls/views.

biodranik reviewed (changes requested):
Thanks! This solution does not fix the core issue, and won't work on all platforms, including iOS.

A proper fix should be done in C++ code that generates these strings. It should be easy now as you already identified the source of these strings.

gpesquero commented:
Thanks @biodranik for your fast response... Your comment makes a lot of sense, so I will have a look at the C++ code and check how we can implement it.

gpesquero commented:
Hello @biodranik... Checking the C++ code it looks like all the distance/speed strings used for routing and search are generated in the same place: `measurement_util.cpp`.

I've updated the commit with a new approach, changing the `ToStringPrecision()` function and getting the decimal separator char from the Locale struct, which I've also changed.

I've tested these changes with the Android emulator on different languages, and everything seems to work Ok, but I don't know if it would also work on ios.

Again, most probably this solution may not be optimal, so any other proposal is welcomed...

biodranik reviewed (commented):
The bicycle was already invented, did you check `[std::ios_base::imbue](https://en.cppreference.com/w/cpp/io/basic_ios/imbue)` ?

gpesquero commented:
Yes, I checked `imbue`, but I could not get it working (most probably due to my low knowledge/experience with cpp)

The code below does compile, but it crashes....

```
string ToStringPrecision(double d, int pr)
{
  stringstream ss;
  ss.imbue(std::locale("es_ES.UTF8"));
  ss << setprecision(pr) << fixed << d;
  return ss.str();
}
```

biodranik commented:
Yeah, sorry, I forgot that NDK does not support C/C++ locales.

Then we have several possible options:
1. The one you implemented. It creates some overhead though.
2. Store the bool flag for dot (or command) in our C++ locale wrapper (pass it from Android and iOS if it also doesn't support locales), make sure that it re-initializes when the device's locale changes, and use that flag internally in the C++ code for a proper choice.
3. Pass raw values to Android and iOS code and format them there instead of C++.

Let's evaluate and compare pros and cons of all options. Need to test it on iOS too.

gpesquero commented:
Hello again... I've managed to get an iOS dev environment working and I've confirmed that the `ss.imbue(std::locale("es_ES.UTF-8"));` line does work on iOS and it provides localized strings.

For the pros and cons analysis, I think that we shall also take into consideration the implementation of the conversion of the thousands separator.

From my point of view, these are the pros and cons if we want to keep the locale conversion in the CPP side (options 1 & 2):

**Opt. 1&2 pros:**
- Conversion done in a single point (no need to make conversion calls all around the UI for both Android/iOS).
- For iOS, it can be directly done in CPP using std locale calls. In CPP in iOS, thousands separator will also be directly done. 

**Opt. 1&2 cons:**
- Locale does not work for Android NDK. Need to make "manual" conversions (as done in option 1) and differently from iOS CPP code.
- Thousand separator shall also be done "manually" for Android NDK.
- Need to pass system locale to CPP code and store in locale wrapper class.
- Need to implement a method to detect system locales changes and update info in wrapper class (Is this really necessary? A system locale change is rarely done, so if the user has to re-start the app in this situation, I don't think that I'd be a big issue...) 

And these are the pros and cons for option 3 (perform conversion in Android/iOS side):

**Opt. 3 pros:**
- No need to store system locale in CPP wrapper class.
- System locale changes are automatically detected.

**Opt. 3 cons:**
- Need to convert strings in many different places in the Android/iOS code.
- Need to convert strings containing distances and speed received from CPP code in different formats (sometimes these strings contain only numbers, and sometimes they also include units as km, miles, km/h).

Please fell free to add or comment all these pros and cons...

biodranik commented:
Thanks for the analysis! If it works on iOS, we can save some time. By the way, did you test it on iOS 12 (the lowest we support)? It is possible to install the simulator and test it.

Another important point IMO is that proper support in C++ will allow it to work properly on all other systems (Linux, Windows, macOS). Otherwise, they must implement the same "conversion" in many places.

There is already some logic to get the current locale/language implemented in platform/preferred_languages.cpp that we can reuse/extend.

The ideal solution should have less code and should not have runtime overhead (e.g. a lot of calculations/allocations each second).

gpesquero commented:
Hello @biodranik!! I've pushed a new commit with a draft implementation for the decimal and grouping (thousands) separator for both Android and iOS in the CPP code.

As we confirmed earlier, Android NDK does not support locales, so the separator conversion is done manually extending the Locale struct. The conversion is done in function `FormatDistanceImpl` and only where it's needed in order to optimize performance.

For iOS, I haven't been able to get `imbue()` to work properly with system locale, so I've decided to use the same solution as for Android, extending also the Locale struct.

I have tested this code using both Android & iOS simulators, and it seems to work OK with some languages and system country settings, but further testing might be needed for many other configurations.

The code also compiles for Linux and macOS desktop environments, though it looks like the `FormatDistance` functions are not used in the desktop versions of the app.

I haven't tried to compile/test it on Windows.

biodranik commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
+
+    if (precision)
+    {
+      // Value with decimals. Set locale decimal separator
```
Mac, Linux, and maybe Windows may properly format the string using the system's locale. Can we at least mention it here? If you see that it should not be corrected on Mac, for example, then an `#ifdef` may be needed.

biodranik replied:
We also try to use . at the end of comment sentences. ;-)

gpesquero replied:
I've added `#ifdef` so the decimal and grouping separators and "manually" replaced only on Android OS.
I've also added a comment explaining this.

biodranik reviewed (commented):
Thank you for your efforts, the code is cleaner now IMO.

> For iOS, I haven't been able to get imbue() to work properly with system locale, so I've decided to use the same solution as for Android, extending also the Locale struct.

What exactly did you try?

Did you try to change the locale globally in the main.mm ?

gpesquero commented:
On IOS I've set the locale globally in `main.mm` with the call `std::locale::global(std::locale(""));`, and I've added also the line `ss.imbue(std::locale(""));` in the `ToStringPrecision(double d, int pr)` function, but the decimal and grouping separators do not change according to the system environment settings.

gpesquero commented on platform/locale.mm:
```diff
@@ -7,8 +7,14 @@
-          locale.currencyCode ? [locale.currencyCode UTF8String] : ""};
+          locale.currencyCode ? [locale.currencyCode UTF8String] : "",
+          decimalSeparatorString.size() > 0 ? decimalSeparatorString[0] : '.',
+          groupingSeparatorString.size() > 0 ? groupingSeparatorString[0] : ','};
```
Yes, empty grouping separators are valid, so I will modify the code to support these situations using a `char(0)` as grouping separator.

gpesquero replied:
Implemented use of `char(0)` for empty group separator

gpesquero commented on android/src/app/organicmaps/util/Utils.java:
```diff
@@ -492,6 +493,18 @@ public static String getLanguageCode()
+  @NonNull
+  public static String getDecimalSeparator()
+  {
+    return String.valueOf(DecimalFormatSymbols.getInstance().getDecimalSeparator());
```
I converted both decimal and grouping separators to string so the jni code was the same as for the country and language code. I've already modified the code so instead of string, we use chars.

gpesquero commented on android/src/app/organicmaps/util/Utils.java:
```diff
@@ -492,6 +493,18 @@ public static String getLanguageCode()
+  @NonNull
+  public static String getGroupingSeparator()
+  {
+    return String.valueOf(DecimalFormatSymbols.getInstance().getGroupingSeparator());
```
Same as before... already changed code to use char instead of strings...

gpesquero commented on platform/locale.hpp:
```diff
@@ -10,6 +10,8 @@ struct Locale
   std::string m_country;
   std::string m_currency;
+  char m_decimal_separator;
+  char m_grouping_separator;
```
Corrected !!

gpesquero commented on platform/locale_linux.cpp:
```diff
@@ -4,7 +4,7 @@ namespace platform
 Locale GetCurrentLocale()
 {
-  return {"", "", ""};
+  return {"", "", "", '.', ','};
```
Replaced "" with {}...

Regarding the Linux code, I don't exactly understand what functionality you want to implement here...

gpesquero commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
     // For distances of 10.0 high units and over round to a whole number, e.g. 9.98 -> 10, 10.9 -> 11
-    return ToStringPrecision(highV, round(highV * 10) / 10 >= 10.0 ? 0 : 1) + " " + high;
+
+    int precision = round(highV * 10) / 10 >= 10.0 ? 0 : 1;
```
Modified !!

gpesquero commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
+      // Value with decimals. Set locale decimal separator
+      if (loc.m_decimal_separator != '.')
+      {
+          std::replace(out.begin(), out.end(), '.', loc.m_decimal_separator);
```
Yes, instead of `std::replace()` we could assume that the decimal separator is at `out.size() - 2`. I will change the code accordingly...
And yes, 3.0 is printed as 3.0

gpesquero replied:
And regarding the unit tests, please confirm that the idea is to add to `measurement_tests.cpp` tests of the `FormatDistanceImpl()` function to check the functionality of the decimal and grouping separators replacement based on different locales.

gpesquero commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
+    {
+      // Value with decimals. Set locale decimal separator
+      if (loc.m_decimal_separator != '.')
+      {
```
Corrected !!

gpesquero commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
+    else
+    {
+      // Value with no decimals. Check if it's equal or bigger than 1000 to insert
+      // the grouping (thousands) separator character
```
The added `#ifdef` applies for both decimal and grouping separator...

gpesquero commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
+          out.insert(out.size()-3, 1, loc.m_grouping_separator);
+        }
+      }
+    }
```
Yes, I can change the code so we can handle two or mode grouping separators...

gpesquero replied:
Done!! I've also moved this piece of code to the function `ToStringPrecision()` so speed strings are also localized...

gpesquero commented on platform/measurement_utils.cpp:
```diff
@@ -49,7 +53,37 @@ string FormatDistanceImpl(Units units, double m, string const & low, string cons
+    // Add units to string
+    out += " " + high;
+
+    return out;
```
Modified !!

gpesquero commented on platform/locale.mm:
```diff
@@ -7,8 +7,14 @@
 {
   NSLocale * locale = [NSLocale currentLocale];
+
+  std::string decimalSeparatorString = [locale.decimalSeparator UTF8String];
```
Modified code to use NSString directly...

gpesquero commented:
One of the checks on macOS of my last commit has failed... I don't know the reason why...

gpesquero commented:
Updated iOS implementation, so now `imbue()` is used with locale string based on language and country codes. Tested successfully with different iOS language settings.

biodranik commented on android/jni/app/organicmaps/platform/Language.cpp:
```diff
@@ -50,8 +50,17 @@ Locale GetCurrentLocale()
                                                                     "()Ljava/lang/String;");
   jni::ScopedLocalRef currencyCode(env, env->CallStaticObjectMethod(g_utilsClazz, getCurrencyCodeId));
 
+  static jmethodID const getDecimalSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getDecimalSeparator",
+                                                                    "()C");
```
```suggestion
  static jmethodID const getDecimalSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getDecimalSeparator", "()C");
```

gpesquero replied:
OK, I've changed the code accordingly.

biodranik commented on android/jni/app/organicmaps/platform/Language.cpp:
```diff
@@ -50,8 +50,17 @@ Locale GetCurrentLocale()
 
+  static jmethodID const getDecimalSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getDecimalSeparator",
+                                                                    "()C");
+  char decimalSeparatorChar = env->CallStaticCharMethod(g_utilsClazz, getDecimalSeparatorId);
```
```suggestion
  char const decimalSeparator = env->CallStaticCharMethod(g_utilsClazz, getDecimalSeparatorId);
```

gpesquero replied:
OK, I've changed the code accordingly.

biodranik commented on android/jni/app/organicmaps/platform/Language.cpp:
```diff
@@ -50,8 +50,17 @@ Locale GetCurrentLocale()
+                                                                    "()C");
+  char decimalSeparatorChar = env->CallStaticCharMethod(g_utilsClazz, getDecimalSeparatorId);
+
+  static jmethodID const getGroupingSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getGroupingSeparator",
+                                                                    "()C");
```
```suggestion
  static jmethodID const getGroupingSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getGroupingSeparator", "()C");
```

gpesquero replied:
OK, I've changed the code accordingly.

biodranik commented on android/jni/app/organicmaps/platform/Language.cpp:
```diff
@@ -50,8 +50,17 @@ Locale GetCurrentLocale()
+
+  static jmethodID const getGroupingSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getGroupingSeparator",
+                                                                    "()C");
+  char groupingSeparatorChar = env->CallStaticCharMethod(g_utilsClazz, getGroupingSeparatorId);;
```
```suggestion
  char const groupingSeparator = env->CallStaticCharMethod(g_utilsClazz, getGroupingSeparatorId);;
```

gpesquero replied:
OK, I've changed the code accordingly.

biodranik commented on iphone/Maps/main.mm:
```diff
@@ -3,11 +3,17 @@
   auto & p = GetPlatform();
   LOG(LINFO, ("Organic Maps", p.Version(), "started, detected CPU cores:", p.CpuCores()));
 
+#ifdef OMIM_OS_IPHONE
```
Doesn't it work on a mac? If it doesn't, then ifdef should also be above near the include. If it does, then ifdef is not needed.

gpesquero replied:
I first assumed that the `std::locale::global(std::locale(""));` call in `main.mm` configured the locale, so certain distance values were displayed with local decimal separator, for example in the place info panel:

![place_info](https://user-images.githubusercontent.com/25888296/219958694-dee50bed-0bc4-4d08-b70b-7a875ef4d278.png)

But now I've realized that these values are correctly displayed with the locale decimal separator because of the use of [MKDistanceFormatter](https://developer.apple.com/documentation/mapkit/mkdistanceformatter) that applies the localized decimal separator.

So this call to `std::locale::global(std::locale(""));` in `main.mm` can be deleted, as it doesn't make any use.

biodranik commented on platform/measurement_utils.cpp:
```diff
@@ -10,25 +10,59 @@
+
+#ifdef OMIM_OS_IPHONE
+  static string const localeString = GetCurrentLocaleString();
+  ss.imbue(std::locale(localeString));
```
Can it be done in main.mm?

biodranik reviewed (commented):
Can you please add a log and track how often is it called in some common scenarios (like searching, for example)?

gpesquero commented:
I've included some logging with the count of calls to the `FormatValueString()` function in the Android version using the simulator and here are the values that I've obtained:

 - Running a search: 90 calls aprox. per search
 - Clicking on a place: around 20 calls per second (from function Framework::GetDistanceAndAzimut)
 - While routing: around 60 calls per second (called 3 times to format total distance, speed and next turn distance)

This function is indeed called quite frequently...

gpesquero commented:
Commit updated to support new `Distance` class, as well as speed decimal separator in iOS.

biodranik reviewed (commented):
Am I correctly understanding that the distance is now properly formatted for each locale, but speed is not?

@AndrewShkrob Does it make sense to extend your distance class with speed support? Or would it be easier to support native speed formatting as it is done right now? Do we need to display speed values from C++ code now (or in the future)?

gpesquero commented:
With this latest commit, both distance **and** speed are formatted on Android and iOS based on system locale.

On iOS, the `Measure` class is used for speed formatting, which is an additional way of dealing with formatting. IMO, the best would be to create a common native `Speed` class, as done for distances.

AndrewShkrob commented:
> Does it make sense to extend your distance class with speed support?

I think, it's better to implement separate `Speed` class.

> Or would it be easier to support native speed formatting as it is done right now?

It's easier with native formatting because it's already implemented. 

> Do we need to display speed values from C++ code now (or in the future)?

Seems that we'll need it at least in the future but it's really hard to make it work in Android (as discussed earlier).

Anyway, I'd prefer formatting directly in C++ more. But as a temporary solution this looks good.

Btw, there is a c++ code that generates info for place page with altitudes formatting that may not be covered with this PR. But I'm not sure

biodranik commented:
> I think, it's better to implement separate Speed class.
> It's easier with native formatting because it's already implemented.

@AndrewShkrob Can you please mention pros and cons of writing a separate Speed class?

Altitude formatting for miles/feets/km should be useful too (but without decimal separator).

@gpesquero what is the overhead in the current implementation?

gpesquero commented:
Hi @biodranik!!

I'm not sure how you usually calculate/estimate overheads...

In any case, I've made some execution time measurements in the function `Distance::GetDistanceString()`, with both implementations running in the same call.

I've used `std::chrono::high_resolution_clock` for the measurements, on a real Android device with USB debugging and logging the results in Android Studio:

Column 1: the current implementation (no locale).
Column 2: new implementation with locale formatting
Column 3: ratio (column 2 divided by column 1)

  50077 ns,  462346 ns, x9
  50192 ns,  414884 ns, x8
  47924 ns,  469692 ns, x10
  30462 ns,  442270 ns, x15
  55192 ns,  441077 ns, x8
  40384 ns,  391346 ns, x10
  48615 ns,  402693 ns, x8
  44962 ns,  390615 ns, x9

This is just a small representative list of the measured values. As you can see, the new implementation takes around 10 times more than the current one.

If we assume 50 us for the typical standard calls (without locale formatting), and around 500 us for the locale formatting implementation, that means a 450 us overhead for each call.

During navigation, these calls are run 20 times per second, and for 3 values (total remaining distance, speed and next turn distance) so total overhead is around 27 ms (20 x 3 x 450 us) per second.

I don't know how this overhead can affect the overall performance of OM.

If you need further testing/measurements, please let me know...

gpesquero commented:
> Btw, there is a c++ code that generates info for place page with altitudes formatting that may not be covered with this PR. But I'm not sure

@AndrewShkrob: The altitude values in the place page are covered in this PR, as you can see here:

![imagen](https://github.com/organicmaps/organicmaps/assets/25888296/144aea5f-64bf-4497-8795-d78ea045eaa9)

But in the planning page of the walking or cycling routes the *ascent* and *descent* values are not formatted based on locale: 

![imagen](https://github.com/organicmaps/organicmaps/assets/25888296/cc416e52-79d0-44d7-8ec6-f46c4667e0c3)

I will have a look at the code to check how it can be also added to this PR for both Android and iOS...

biodranik commented:
10x slow down for an often used operation is not good. 

Can we consider again a proper formatting in C++ where possible, and proper updating of system locale changes in C++ code? Or is it possible to do everything exclusively on a platform side, without involving C++?

gpesquero commented:
I agree with you that such a big overhead for this often and minor functionality is not acceptable.

Taking into account all the analysis and tests that we've done so far in this PR and in issue [3992](https://github.com/organicmaps/organicmaps/issues/3992), I think that the best approach from performance and code compatibility across platforms PoV is to perform locale formatting in C++ using custom manual decimal and grouping separators replacement on the strings generated in C++.

This solution can seem to be rather simple and dumb, but doing the formatting  in the platform specific side would increase overhead and the need to duplicate the code. 

This simple solution may not deal with 100% of the possible locales, such as the Lakh Indian numbering system, but I assume that this numbering system would not apply to the display of the often calls in navigation mode, and we could always implement these other numbering systems manually if needed.

One possible solution of compromise would be two have to different locale formatting functions: one for fast/often locale formatting in navigation mode (speed, next turn and remaining distance), and another one for standard/not so often calls (search results, place info panel, etc.).

I can make a proposal of dual fast/often & standard/seldom locale formatting calls, and see how it works/looks...

gpesquero commented:
Updated commit with a proposal of dual locale formatting:

- Fast & often locale formatting, using manual replacement in c++ of the decimal and grouping separators in function `ToStringPrecision()`. This function is only used for fast calls during navigation, adding a minimum overhead.

- Standard & seldom locale formatting, using Android & iOS platform specific calls. These calls are used for the rest of functions (search results, place info panel, etc.) which are not used very often.

The selection between fast or standard mode is done in the function  `Distance::GetDistanceString(bool fastMode = false)`, where the standard mode is used as default.

This commit has only been tested (for the moment) in Android.

I'm not really convinced of this dual locale formatting. I'd rather keep only the fast & often implementation in c++, which would make the code clearer, we would have always a better performance, and we will support 99% of the possible locales. If in the future we have to implement a different locale formatting, we could do it in c++, which may not be really difficult.

biodranik commented:
> I'm not really convinced of this dual locale formatting. I'd rather keep only the fast & often implementation in c++, which would make the code clearer, we would have always a better performance, and we will support 99% of the possible locales. If in the future we have to implement a different locale formatting, we could do it in c++, which may not be really difficult.

I agree with you, these are all good points.

gpesquero commented:
I've removed the '_dual formatting_', so now we only have the fast locale formatting in c++ for the decimal and grouping separators.

While making some tests in different locales in both Android and iOS, and I've noticed that some languages (for example, French) use for the grouping separator non-standard characters such as the NNBSP (NARROW NO-BREAK SPACE, U+202F) , which (AFAIK) cannot be handled with a `char` variable in c++.

I've tried to use `wchar_t` variables, but their insertion in `std::string` is not direct, so I've finally decided instead to use `std::string` variables to store the decimal and grouping separator characters, which I think it does not add a big overhead.

biodranik commented on android/jni/app/organicmaps/Framework.cpp:
```diff
@@ -1294,6 +1295,12 @@ Java_app_organicmaps_Framework_nativeGenerateRouteAltitudeChartBits(JNIEnv * env
     totalDescent = measurement_utils::MetersToFeet(totalDescent);
   }
 
+  jni::TScopedLocalRef const totalAscentString(env,
+                             jni::ToJavaString(env, ToStringPrecision(totalAscent, 0)));
+
+  jni::TScopedLocalRef const totalDescentString(env,
+                             jni::ToJavaString(env, ToStringPrecision(totalDescent, 0)));
```
```suggestion
  jni::TScopedLocalRef const totalAscentString(env, jni::ToJavaString(env, ToStringPrecision(totalAscent, 0)));

  jni::TScopedLocalRef const totalDescentString(env, jni::ToJavaString(env, ToStringPrecision(totalDescent, 0)));
```

gpesquero replied:
Noted! Commit will be amended accordingly...

biodranik commented on android/jni/app/organicmaps/platform/Language.cpp:
```diff
@@ -50,8 +50,18 @@ Locale GetCurrentLocale()
   jni::ScopedLocalRef currencyCode(env, env->CallStaticObjectMethod(g_utilsClazz, getCurrencyCodeId));
 
+  static jmethodID const getDecimalSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getDecimalSeparator",
+                                                                    "()Ljava/lang/String;");
```
Align it with (

gpesquero replied:
Noted! Commit will be amended accordingly...

biodranik commented on android/jni/app/organicmaps/platform/Language.cpp:
```diff
@@ -50,8 +50,18 @@ Locale GetCurrentLocale()
+  jni::ScopedLocalRef decimalSeparatorChar(env,  env->CallStaticObjectMethod(g_utilsClazz, getDecimalSeparatorId));
+
+  static jmethodID const getGroupingSeparatorId = jni::GetStaticMethodID(env, g_utilsClazz, "getGroupingSeparator",
+                                                                    "()Ljava/lang/String;");
```
ditto

gpesquero replied:
Noted! Commit will be amended accordingly...

biodranik commented on iphone/CoreApi/CoreApi/Common/MWMGeoUtil.mm:
```diff
@@ -31,10 +32,8 @@ @implementation Measure
-  else
-    return [NSString stringWithFormat:@"%.1f", self.value];
+  auto outString = measurement_utils::ToStringPrecision(self.value, self.value >= 10.0 ? 0 : 1);
+  return [NSString stringWithCString:outString.c_str() encoding:[NSString defaultCStringEncoding]];
```
stringWithUTF8String

gpesquero replied:
Noted! I will use `stringWithUTF8String` instead of `stringWithCString`.

biodranik commented on iphone/CoreApi/CoreApi/Common/MWMGeoUtil.mm:
```diff
@@ -31,10 +32,8 @@ @implementation Measure
-    return [NSString stringWithFormat:@"%.0f", self.value];
-  else
-    return [NSString stringWithFormat:@"%.1f", self.value];
+  auto outString = measurement_utils::ToStringPrecision(self.value, self.value >= 10.0 ? 0 : 1);
```
```suggestion
  auto const outString = measurement_utils::ToStringPrecision(self.value, self.value >= 10.0 ? 0 : 1);
```

gpesquero replied:
Noted! Commit will be amended accordingly...

biodranik commented on platform/locale.hpp:
```diff
@@ -10,6 +10,8 @@ struct Locale
   std::string m_language;
   std::string m_country;
   std::string m_currency;
+  std::string m_decimalSeparator;
```
Can a decimal separator also be a non-ASCII symbol? Is there a full list somewhere?

gpesquero replied:
Yes, according to [this](https://en.wikipedia.org/wiki/Decimal_separator) Wikipedia article, there's also an Arabic decimal separator (U+066B).

Here's a screenshot of the place info panel with locate set to Farsi/Persian, where the decimal and grouping separators are both non-ASCII chars:
![Captura desde 2023-08-17 00-08-14](https://github.com/organicmaps/organicmaps/assets/25888296/cceb989c-00ed-435b-9c64-a33706e18622)

biodranik commented on platform/platform_tests/distance_tests.cpp:
```diff
@@ -282,19 +282,19 @@ UNIT_TEST(Distance_FormattedDistance)
+    {Distance(999,     Units::Miles),      999,  Units::Miles,      "999",   "999 mi"},
+    {Distance(1149.99, Units::Miles),      1150, Units::Miles,      "1,150", "1,150 mi"},
+    {Distance(1150,    Units::Miles),      1150, Units::Miles,      "1,150", "1,150 mi"},
+    {Distance(1150.01, Units::Miles),      1150, Units::Miles,      "1,150", "1,150 mi"},
```
1. Can tests for a decimal separator also be added?
2. Is it hard to add tests for altitude formatting?
3. What if tests are run on a French locale?

gpesquero replied:
> 1. Can tests for a decimal separator also be added?

Do you mean to test decimal separators on different locales? With the current implementation in `ToStringPrecision()`, `GetCurrentLocale()` is only called once at start-up, so we could only make tests with a single locale.

> 2. Is it hard to add tests for altitude formatting?

We already have a couple of tests for altitude formatting in `UNIT_TEST(Distance_CreateAltitudeFormatted)` in `distance_tests.cpp`. I've included tough a third one for high altitudes in meters.

> 3. What if tests are run on a French locale?

Current implementation of `GetCurrentLocale()` in linux does not read system's locale, and forces decimal and grouping separators to '.' and ',' respectively, so tests run flawlessly even with Spanish or French system locales.

biodranik replied:
1/ I meant test cases where a decimal separator is used.

biodranik reviewed (commented):
@AndrewShkrob @vng WDYT?

biodranik commented:
BTW, @gpesquero thank you for your time and efforts to improve Organic Maps! I value and respect your efforts a lot.

AndrewShkrob commented on platform/locale.mm:
```diff
@@ -9,6 +9,8 @@ Locale GetCurrentLocale()
           locale.countryCode ? [locale.countryCode UTF8String] : "",
-          locale.currencyCode ? [locale.currencyCode UTF8String] : ""};
+          locale.currencyCode ? [locale.currencyCode UTF8String] : "",
+          locale.decimalSeparator ? [locale.decimalSeparator UTF8String] : ".",
+          locale.groupingSeparator ? [locale.groupingSeparator UTF8String] : ","};
```
I'm not sure that this default value is OK. I think, it's better to leave it empty.

gpesquero replied:
Yes, maybe an empty grouping separator may be more common across different locales, though this default value will be overwritten by system's configuration in most of the times.

Your comment also confirms that in many languages, the grouping separator is only used in values with more than 4 digits, keeping digits with 4 digits with no grouping separator ("10 123 m" and "1234 m"). Maybe we shall also keep this behavior, which is also _suggested_ as International Standard according to [this](https://en.wikipedia.org/wiki/Decimal_separator#Digit_grouping) Wikipedia article about digit grouping. This International Standard also _suggests_ to use always the narrow no-break space char as the grouping separator.

My proposal for the grouping separator would be then: set empty "" (no grouping separator) as default if the system does not provide any & apply grouping separators *only* on numbers with more that 4 digits.

biodranik replied:
Thanks! Looks like a good default is:
- Use group separator for >4 chars only
- Use a non-breaking space

P.S. Is non-breaking space used after a number and before the unit name? Should we use it too?

gpesquero replied:
@biodranik: I assume that you suggest to use as default the **narrow** non-breaking space (NNBSP, U+202F), instead of the **standard** non-breaking space (NBSP, U+00A0).

In [this](https://en.wikipedia.org/wiki/Non-breaking_space#Uses_and_variations) Wikipedia article, it's indeed stated that the NBSP's are also used between the value and the units, so yes, we can also use them there instead of the regular whitespace.

In a first commit I've set the NNBSP as default for iOS and linux and I've modified the UTs accordingly, but after pushing the commit, the iOS tests have failed because the grouping separator is configured to the ',' character. The workaround is to set the default character in linux and the UTs with the ',' character.

I've amended the commit according to these last comments.

AndrewShkrob commented on platform/locale_linux.cpp:
```diff
@@ -4,7 +4,7 @@ namespace platform
 Locale GetCurrentLocale()
 {
-  return {"", "", ""};
+  return {{}, {}, {}, ".", ","};
```
Same here

gpesquero replied:
Same as before, I agree to set the empty "" grouping separator as default, though this change affects the unit testing code, so I will have to change back again the unit tests for distances.

AndrewShkrob commented:
LGTM.

@gpesquero It would be great if you can also add UTs for `ToStringPrecision` function to check it's behavior with different locales.

gpesquero commented:
> LGTM.

Thanks @AndrewShkrob for your review!!

> @gpesquero It would be great if you can also add UTs for `ToStringPrecision` function to check it's behavior with different locales.

As I've stated in a previous comment, the current implementation of the `ToStringPrecision()` function only calls `GetCurrentLocale()` once at start-up, which (AFAIK) does not allow to make tests with different locales.

Any suggestion how to implement these unit tests is welcomed!!

biodranik commented on platform/platform_tests/distance_tests.cpp:
```diff
@@ -256,21 +265,21 @@ UNIT_TEST(Distance_FormattedDistance)
+    {Distance(100.01,  Units::Kilometers), 100,  Units::Kilometers, "100",   "100 km"},
+    {Distance(115,     Units::Kilometers), 115,  Units::Kilometers, "115",   "115 km"},
+    {Distance(999,     Units::Kilometers), 999,  Units::Kilometers, "999",   "999 km"},
+    {Distance(1000,    Units::Kilometers), 1000, Units::Kilometers, "1,000", "1,000 km"},
```
If we change defaults to >4 chars for group separators, these tests will change, right?

gpesquero replied:
Yes, the UTs have to be changed.

biodranik commented on android/src/app/organicmaps/routing/RoutingBottomMenuController.java:
```diff
@@ -265,9 +265,8 @@ private void showRouteAltitudeChart()
-      mAltitudeDifference.setText(String.format(Locale.getDefault(), "↗ %d %s ↘ %d %s",
-                                                limits.totalAscent, unit,
-                                                limits.totalDescent, unit));
+      mAltitudeDifference.setText("↗ " + limits.totalAscentString + " " + unit +
```
Is + faster/better than String.format?

gpesquero replied:
Yes, the + operator seems to be much faster:

First column: String.format, second column: + operator
2778346 ns, 72961 ns
1656885 ns, 341115 ns
1513307 ns, 45154 ns
1233769 ns, 32039 ns

And as we're dealing now only with strings, the code seems to be clearer than with the String.format function.

biodranik reviewed (commented):
Good! @Jean-BaptisteC can you please test if it works with different system languages, including Arabic, Hebrew, French, German, US, Chinese, Japanese?

Jean-BaptisteC reviewed (approved):
✅ Pixel 6 - Android 13

French
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/a0717cfe-ebe5-41c7-8612-f9da8f95d892" height=50 />
Arabic
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/30ab24a6-bb5b-45f9-acb0-5babff9c0f7b" height=50 />
Deutsch
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/be4674dd-162d-4368-bfe9-9c4e53f7769f" height=50 />
English (US)
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/b9f8c647-2d4a-4933-8170-ef4138a54c89" height=50 />
Japanese
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/0cf201da-62c9-4368-a6cf-393945f0f523" height=50 />
Chinese
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/bc29a30c-55e9-4f6e-8d74-4db97379c074" height=50 />
Hebrew
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/ec29aadc-48bc-435f-81c1-865432a97179" height=50 />

Other examples:
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/a4ad6862-41d3-4386-86b8-a8f1469fc015" height=60 />

<img src="https://github.com/organicmaps/organicmaps/assets/87148630/5114c5d8-e844-4ae5-aeb3-2617000b3012" height=350 />

<img src="https://github.com/organicmaps/organicmaps/assets/87148630/8abbfef9-4204-49a7-840d-ee9a362bd6b8" height=350 />

gpesquero commented:
> @gpesquero It would be great if you can also add UTs for `ToStringPrecision` function to check it's behavior with different locales.

Hi @AndrewShkrob: I've implemented UTs for the `ToStringPrecision()` function working on different locales, by creating an additional function called `ToStringPrecisionLocale()` which takes as argument the current locale.

In order to run the new tests locally on my Linux machine, I had to install missing locales as for french using the following commands:

`sudo locale-gen fr_FR`
`sudo locale-gen fr_FR.UTF-8`
`sudo update-locale`

Now I've pushed the commit with these new tests, but the Linux checks which include tests fail due to missing locales.

Does anyhow know how to setup these locales in the GitHub checks?

biodranik reviewed (approved):
Doing _even faster_ string generation without iostream and reallocations would be an awesome bonus, but it's good enough already.

@vng WDYT?

biodranik reviewed (changes requested):
Oh wait:

```
1  tests failed:
distance_tests.cpp::Distance_ToStringPrecisionLocale
Some tests FAILED.
```

AndrewShkrob commented on platform/platform_tests/distance_tests.cpp:
```diff
@@ -238,97 +256,99 @@ UNIT_TEST(Distance_FormattedDistance)
-    {Distance(633'600, Units::Feet),       120,  Units::Miles,      "120",  "120 mi"},
-    {Distance(633'653, Units::Feet),       120,  Units::Miles,      "120",  "120 mi"},
-    {Distance(999'999, Units::Feet),       189,  Units::Miles,      "189",  "189 mi"},
+    {Distance(999,     Units::Feet),       0.2,   Units::Miles,      "0.2",    "0.2\u202Fmi"},
```
Let's move `\u202F` to a const variable at the beginning of the file:
`char const kNonBreakingSpace = '\u202F'`

biodranik replied:
constexpr is even better.

gpesquero replied:
As the non-breaking space `\u202F` is used in different source files, I've included its definition in the file `locale.hpp`.

I cannot get it working as `char`, so I've declared it as `std::string`.

AndrewShkrob commented on platform/platform_tests/distance_tests.cpp:
```diff
@@ -346,4 +366,34 @@ UNIT_TEST(Distance_FormattedDistance)
   }
 }
 
+UNIT_TEST(Distance_ToStringPrecisionLocale)
```
1. Let's move it to the related tests file. I guess, it's `measurement_tests.cpp`
2. I don't really like this solution.
   You can mock locale as it's done for display units with `ScopedSettings guard`.
   Create `ScopedLocale guard` class, pass separators to the constructor and overwrite current locale settings. In destructor - restore locale.
   Check `UNIT_TEST(Distance_CreateFormatted)` as an example

This solution won't require installing additional locales to the system.

biodranik replied:
Testing specific locales in a controlled environment is ok. You can edit our .github/workflows actions by adding `sudo locale-gen ru_RU.UTF-8` commands in the beginning.

gpesquero replied:
I've added the `sudo locale-gen xx_YY.UTF-8` commands to `.github/workflows/linux_check.yaml` for different languages (en, es, fr & ru) and now the linux checks seem to work Ok.

Anyhow, the macOS checks still do not work. Instead of the linux command `locale-gen`, in macOS it looks like the command to use is `localedef`, but I still cannot get it working. I will make further investigations in the incoming days, but any help is welcomed.

AndrewShkrob replied:
That will definetly break tests on local machines without all required locales.
At least, consider covering such test cases with `#if CI`
```cpp
    TestData testData[] = {
          { "en_US.UTF-8",     "9.8",      "12,345"},
#if CI
          { "es_ES.UTF-8",     "9,8",      "12.345"},
          { "fr_FR.UTF-8",     "9,8",      "12 345"},
          { "ru_RU.UTF-8",     "9,8",      "12 345"}
#endif
    };
```

We can pass `CI` var through cmake when it's defined in the environment. CI services like GitHub Actions always have this environment variable.

gpesquero replied:
Let me describe the latest tests and implementations that I've perfomed today:

- Finally it was not needed to use the command `localedef` in macOS. I only had to rename the locales from `xx_YY.UTF8` to `xx_YY.UTF-8` to get them working in macOS.
- I found out that in macOS the call to `std::locale::global()` does not affect the behavior of class `NSLocale`, so I've had to implement a new call `GetLocale(std::string)` in `locale.mm` order to obtain a new Locale struct based on the locale name in macOS.
- Now comes the funny part: I've made tests with 4 different locales: EN, ES, FR and RU. With EN and ES there are no problems, but with FR locale the grouping separator character in Linux is a normal whitespace (\u0020), but in macOS it's a Narrow Non-Breaking Space (NNBSP, \u202F). And with RU locale, in Linux it's also a normal whitespace (\u0020), but in macOS it's a standard Non-Breaking Space (NBSP, \u00A0).

There's a mix of many different types of space characters in some locales and in different platforms, and the only possible work-around is to fill the testing code with `#ifdef` for the different platforms and locales.

Taking into consideration all these tests and the comment from @AndrewShkrob...

> That will definetly break tests on local machines without all required locales.
> At least, consider covering such test cases with #if CI

... I think that it's no worth the implementation of the UTs for the `ToStringPrecision()` function with different locales.

Let me know your opinion whether we shall continue to implement these UTs.

gpesquero replied:
Finally I decided to remove the UTs for the `ToStringPrecision()` function with different locales.

biodranik replied:
Testing it on different OSes with different locales actually was quite useful )

gpesquero replied:
Ok, I've brought back the UTs for the `ToStringPrecision()` function with different locales.

As the grouping separator is different for Linux and MacOS in some languages, the `UNIT_TEST(ToStringPrecisionLocale)` contains `#ifdefs` for different test results in different OS's.

I've included also a `#if CI` as suggested by @AndrewShkrob, so these tests are only performed in the GitHub workflows, and local tests do not require locales to be installed.

AndrewShkrob commented on .github/workflows/linux-check.yaml:
```diff
@@ -163,5 +163,14 @@ jobs:
           # world_feed_integration_tests - https://github.com/organicmaps/organicmaps/issues/215
           CTEST_EXCLUDE_REGEX: "drape_tests|generator_integration_tests|opening_hours_integration_tests|opening_hours_supported_features_tests|routing_benchmarks|routing_integration_tests|routing_quality_tests|search_quality_tests|storage_integration_tests|shaders_tests|world_feed_integration_tests"
         run: |
+          sudo locale-gen en_US
+          sudo locale-gen en_US.UTF-8
+          sudo locale-gen es_ES
+          sudo locale-gen es_ES.UTF-8
+          sudo locale-gen fr_FR
+          sudo locale-gen fr_FR.UTF-8
+          sudo locale-gen ru_RU
+          sudo locale-gen ru_RU.UTF-8
+          sudo update-locale
```
I think it's better to move this part to a separate job. Probably near the dependencies job or before the tests job with a comment that it's required only for testing

gpesquero replied:
Hi @AndrewShkrob! I've removed the UTs for the `ToStringPrecision()` function, so these calls to `sudo locale-gen XX_YY` are no longer needed. Thanks anyway for your comment.

gpesquero replied:
Hello again @AndrewShkrob! As I've included back again the UTs for the  `ToStringPrecision()` function, I've included the `sudo locale-gen xx_YY` calls in the Github workflow for Linux, in the same place that I initially used.

You suggested to move this part to a separate job, but I'm not sure how to do it.

Currently, Linux tests are run in the last step 'Tests' of the 'linux-matrix' job.

I'd appreciate if you could provide a piece of code of your proposal for this change.

biodranik replied:
1. You can move these calls to the tests section. 
2. Is there a way to detect if locale is installed and skip the test if it's not?
3. Is there a way to get a locale-specific delimiter from a test and avoid ifdef?

gpesquero replied:
1. I still don't know where you want these calls to be placed. They already are located in the 'Tests' step of the 'linux-matrix' job, and I don't see any other tests section in this 'linux-check.yaml" file. Please provide a diff file or some direct instructions (like: insert these calls in line XX of file YY), otherwise I'm lost regarding this change request.

2. I've added some try/catch in the `GetLocale()` function in Linux to detect errors while loading the locales, so now the tests don't crash and we can skip the tests for the missing locales. Under macOS, `GetLocale()` never fails, even if a wrong locale is requested.

3. I've removed the `#ifdefs` for grouping separator checking in different OS's by adding a `AddGroupingSeparators()` function that uses locale's grouping separator for the tests.

AndrewShkrob commented on platform/platform_tests/distance_tests.cpp:
```diff
@@ -45,7 +45,7 @@ UNIT_TEST(Distance_CreateFormatted)
     TEST_ALMOST_EQUAL_ULPS(d.GetDistance(), 100.0, ());
     TEST_EQUAL(d.GetDistanceString(), "100", ());
-    TEST_EQUAL(d.ToString(), "100 m", ());
+    TEST_EQUAL(d.ToString(), "100" + kNonBreakingSpace + "m", ());
```
`"100" + kNonBreakingSpace + "m"`

I apologize for raising what might seem like trivial matters, but I believe it would enhance code readability to establish a distinct function. This way, we can encapsulate the use of kNonBreakingSpace and prevent its repetition on every line.

gpesquero replied:
Hi AndrewShkrob! Please do not hesitate to make any comment about my code. You're most probably used to the coding standards/style in OM.

I've added the `static` function `Distance::ToString()` which encapsulates the usage of `kNonBreakingSpace`.

Please have a look at the code that I've just pushed and confirm that this approach is Ok.

AndrewShkrob replied:
Certainly, there's no need for such a function inside the Distance class. Since it will solely be utilized in unit tests, I suggest moving it to `platform/platform_tests/distance_tests.cpp`.

Moreover, I concur that `ToString` might not be the most suitable name. How about renaming it to `MakeDistanceStr` or a similar variant?

gpesquero replied:
Ok!! I've moved the function (now called `MakeDistanceStr()`) to `platform/platform_tests/distance_tests.cpp`.

gpesquero commented:
During the past few hour,s all the GitHub checks fail due to the same failure:

`fatal: unable to access 'https://gitlab.freedesktop.org/freetype/freetype.git/': The requested URL returned error: 504`
`fatal: clone of 'https://gitlab.freedesktop.org/freetype/freetype.git' into submodule path '/Users/runner/work/organicmaps/organicmaps/3party/freetype/freetype' failed`
`Failed to clone '3party/freetype/freetype' a second time, aborting`
`Error: Process completed with exit code 1.`

Any idea why this is happening ??

biodranik commented:
> Any idea why this is happening ??

https://github.com/organicmaps/organicmaps/pull/5848 should fix it. Please rebase after it is merged.

AndrewShkrob commented on platform/platform_tests/distance_tests.cpp:
```diff
@@ -6,6 +6,11 @@
 
 namespace platform
 {
+std::string MakeDistanceStr(std::string value, std::string unit)
```
```suggestion
std::string MakeDistanceStr(std::string const & value, std::string const & unit)
```

biodranik commented on platform/locale_linux.cpp:
```diff
@@ -1,10 +1,48 @@
+  try
+  {
+    // Set global locale
+    std::locale::global(std::locale(localeName));
```
Why setting a global locale required? Does `std::locale(localeName)` work directly?

gpesquero replied:
You're right! There's no need to use `std::locale::global()`. Will change accordingly...

gpesquero replied:
Commit amended!!

biodranik reviewed (approved):
Thank you very much for your work! And sorry for the delay. Let's see how it works in production :)

gpesquero commented:
Thanks @biodranik! I've been testing it with the Beta version and (for the moment), seems to work fine...

biodranik commented:
@gpesquero user reported this bug, can it be related?

> After an upgrade to the latest version, in the navigation mode, the distance to the next turn or to the final destination freezes at 10 kilometers (when the countdown should switch from counting kilometers to counting hundreds of meters) and unfreezes back only when the distance gets below 1 kilometer (when it switches to counting dozens of meters).

gpesquero commented:
> @gpesquero user reported this bug for iOS, can it be related?

I'm away now from my running/development iOS environment. I won't be able to test and try to replicate this situation till this weekend. As soon as I make the tests, I will let you know...

gpesquero commented:
> @gpesquero user reported this bug for iOS, can it be related?

My iOS development/testing environment is quite limited, so I haven't been able to reproduce this incorrect behavior. I cannot confirm whether the setting of the decimal separator has anything to do with it...

biodranik commented:
@gpesquero tests are failing on Mac with en_CH locale.
```
Running distance_tests.cpp::Distance_CreateAltitudeFormatted
FAILED
platform_tests/distance_tests.cpp:92 TEST(d.GetDistanceString() == "12,345") 12'345 12,345 
Test took 0 ms

Running distance_tests.cpp::Distance_FormattedDistance
FAILED
platform_tests/distance_tests.cpp:368 TEST(d.GetDistanceString() == data.formattedDistanceString) 12'345 12,345 12'345 km
Test took 79990 ms
```

gpesquero commented:
> @gpesquero tests are failing on Mac with en_CH locale.

You're right @biodranik... On macOS the tests seem to use system's locale for the value string formatting...

I've launched [PR 6470](https://github.com/organicmaps/organicmaps/pull/6470) to solve this issue, though some of the checks fail... I still need some time to get them fixed...
